### PR TITLE
A tutorial generation script with Fire.

### PR DESCRIPTION
**Pull request by GiggleLiu**

### CODE REVIEW

File: generate.jl [hunk 1]
```diff
@@ -0,0 +1,27 @@
+using Pkg
+Pkg.activate(".")
+using Fire
+using PhysicsTutorials
+
+path_names(path::String) = path in ("","/") ? () : (path_names(dirname(path))..., basename(path))
```
Comment (carstenbauer) [inline generate.jl]: Use [`splitpath`](https://docs.julialang.org/en/latest/base/file/#Base.Filesystem.splitpath).

File: generate.jl [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+@main function generate(filename::String)
+    pns = path_names(abspath(filename))
+    length(pns) < 3 && error("illegal path $filename")
```
Comment (carstenbauer) [inline generate.jl]: "Illegal" sounds a bit strong. Maybe replace by something like `Invalid path to source file. Expected format: "tutorials/<category>/<tutorial_name>/<tutorial_name>.ipynb (or .jl/.jmd)"`?

File: generate.jl [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    fname==tutorial_name || error("tutorial name should be consistent with filename, e.g. `tutorials/quantum_computing/<name>/<name>.ipynb`")
+
+    if ext == ".ipynb"
+        source = PhysicsTutorials.NotebookSource()
+    elseif ext ==".jmd"
+        source = PhysicsTutorials.WeaveSource()
+    elseif ext ==".jl"
+        source = PhysicsTutorials.LiterateSource()
+    else
+        error("illegal extension $filename")
```
Comment (carstenbauer) [inline generate.jl]: See comment above. We could remind the user what source file extensions are supported.

File: generate.jl [hunk 4]
```diff
@@ -0,0 +1,27 @@
+using Pkg
+Pkg.activate(".")
```
Comment (carstenbauer) [inline generate.jl]: Better use `__@DIR__` here to be safe.

File: README.md [hunk 5]
```diff
@@ -29,6 +29,13 @@ PhysicsTutorials.open_notebooks()
 Supported source files for tutorials are Jupyter notebooks, Weave.jl files, or Literate.jl files.
 To contribute a tutorial, clone the repository and put the source file into `tutorials/<category>/<tutorial_name>/` and name it `<tutorial_name>.ipynb` (extension `.jmd`/`.jl` for Weave/Literate sources). To trigger the generation process of all output formats, run the following code from within the repository root folder:
```
Comment (carstenbauer) [inline README.md:30]: We should change the "to trigger the generation process of all ...." by "to trigger the conversion to all other output formats...".

File: README.md [hunk 6]
```diff
@@ -29,6 +29,13 @@ PhysicsTutorials.open_notebooks()
...
 
+0. Make sure you have `Fire` installed, if not, install it with `julia>]add Fire`,
```
Comment (carstenbauer) [inline README.md]: Perhaps simply `] add Fire` to allow for copy and paste?

File: README.md [hunk 7]
```diff
@@ -29,6 +29,13 @@ PhysicsTutorials.open_notebooks()
...
+1. Create source file `tutorials/<category>/<tutorial_name>/<tutorial_name>.<extension>`, where the extension should be one of `.ipynb`, `.jl` and `.jmd` .
+2. Type `julia generate.jl tutorials/<category>/<tutorial_name>/<tutorial_name>.<extension>` in a terminal.
```
Comment (carstenbauer) [inline README.md]: `convert.jl` seems like a better name (see comment above).

File: generate.jl (hunk 1 above)
Comment (carstenbauer) [reply generate.jl]: However, note that it has only been added in v1.1. So perhaps define a custom `splitpath` in a `if VERSION <= v"1.1"` block.

Also, note that your `path_names` function doesn't work on Windows, i.e. try `path_names("C:\\test")`.

File: README.md [hunk 8]
```diff
@@ -29,6 +29,13 @@ PhysicsTutorials.open_notebooks()
...
+
+
+#### An optional approach: generate source manually
```
Comment (carstenbauer) [inline README.md]: Maybe "Alternative approach: trigger conversion manually"?

**Review by carstenbauer — CHANGES_REQUESTED** [review]
Thanks for the separate PR! Let me leave a couple of comments.

File: README.md (hunk 6 above)
Comment (GiggleLiu) [reply README.md]: Here, a space is missing, `julia> ] add Fire` behaves the same as `] add Fire` when one copys it to REPL (julia REPL will skip `julia> ` automatically).
Here, the problem is, both of them do not allow copy and paste?

File: generate.jl (hunk 1 above)
Comment (GiggleLiu) [reply generate.jl]: I was intended to use `path_names` to solve the compatibility issue between Julia 1.0 and Julia 1.1, but didn't test for windows users. Now I copy paste the Julia 1.1 implementation of `splitpath` to `convert.jl`, this works pretty well for both Julia 1.0 and julia 1.1 and hopefully work for Windows users too.

File: README.md (hunk 6 above)
Comment (carstenbauer) [reply README.md]: > Here, the problem is, both of them do not allow copy and paste?

Why? I don't see the problem.

Comment (GiggleLiu) [reply README.md]: On my laptop, it does not enter pkg mode, instead, it prints a `]`. My system is Ubuntu 18.04.

Comment (carstenbauer) [reply README.md]: For me (Windows 10), pasting `] add Fire` switches to pkg mode (desired behavior) while pasting `julia> ] add Fire` does not work:

```julia
julia> julia> ] add Fire
ERROR: syntax: unexpected "]"
```

Anyways, I feel `] add Fire` is our best bet here. I'll merge and change this myself afterwards.

**Comment by carstenbauer** [comment]
Thanks!

**Pull request merged**